src/DotNetEnv/Parsers.cs

```diff
@@ -176,14 +177,16 @@ private static string ToUtf32Char (string hex)
         // unquoted values can have interpolated variables,
         // but only inline whitespace -- until a comment,
         // and no escaped chars, nor byte code chars
-        // FIXME: would be nice to solve multi word with comment case with parser directly (no split + trim)
         internal static readonly Parser<ValueCalculator> UnquotedValue =
-            InterpolatedValue.Or(
-                NotControlNorWhitespace("'\"$")
-                .Or(InlineWhitespaceChars.AtLeastOnce().Text())
-                .AtLeastOnce()
-                .Select(strs => new ValueActual(strs))
-            ).Many().Select(vs => new ValueCalculator(vs).Split("[ \t]#").Trim());
+            Parse.Regex("[ \t\"']").Not()
```

Comment:
Why is this still here?

Reply:
This is about the disallowed leading characters, and assures that we throw if there is a wrong character.
We can't do that inside, because due to the `Many` we would return an empty Value there.
And we need the `Many`, because it is fine to have no value, it is just not allowed to begin with one of these chars.

Reply:
Please walk me through why you think this implementation is better than mine for unquotedvalue.

Reply:
In my opinion the main point is about the "completeness/correctness" of the unquotedValue (I try to reduce it to the core here, long version: https://github.com/tonerdo/dotnet-env/pull/87#issuecomment-1860793258), which results in the following:

1. User-Experience: if there is an unmatched quote, we get the "correct" exceptionmessage now, because unquotedValue will not succeed with "EmptyValue" and fails instead
2. There is a smaller risk of unexpected results if there is a bug within quotedValues, because unquotedValue is proven by Tests to fail if a value starts with a quote
3. So after all, an additional Test which prevents mistakes for any following PR

The main thing I'm not really happy with my own solution is about the `Then`.
If there is a way to make it more obvious why the parameterd in the `Then` are discarded that would be great...
But in general I like the explicit `Not`. I guess will try a solution like with QuotedValues, where the Contents are separated and a `From-Select` is used instead of `Then`.

Reply:
```
        // unquoted values can have interpolated variables,
        // but only inline whitespace -- until a comment,
        // and no escaped chars, nor byte code chars
        private static readonly Parser<ValueCalculator> UnquotedValueContents = InterpolatedValue
                .Or(from inlineWhitespaces in InlineWhitespace
                    from _ in Parse.Char('#').Not()
                    from partOfValue in NotControlNorWhitespace("$\"'")
                    select new ValueActual(string.Concat(inlineWhitespaces, partOfValue)))
                .Many()
                .Select(vs => new ValueCalculator(vs));

        internal static readonly Parser<ValueCalculator> UnquotedValue =
            from _ in Parse.Regex("[ \t\"']").Not()
            from value in UnquotedValueContents
            select value;
```

This way we don't need to use the `Then`, which is better for readability I guess.
Due to the separation made by the `from` statements I would say we see the reason for discarding quite obvious now, equal to the QuotedValues, where we discard the quotes.

Just wanted to have you a look through it before pushing something new.
Tests still succeed, if I'm right we just replace `Then` by `from-select`.

Reply:
My personal interpretation of yours, in contrast to mine: https://github.com/tonerdo/dotnet-env/pull/87/files#r1437281239

Yours is (if I understand it correctly):
1. starts w a regex, but not-ed ("not" is a often confusing move in languages in general, and same in logic, imo) -- I'm not even sure exactly what this is supposed to match, obviously to prevent whitespace and quotes, but can't that be handled more directly by the matchers below? As in, why do you need this at all?
2. Then allow interpolated values -- I'm confused how this doesn't satisfy the initial regex+not to begin with, tbh
3. Or whitespace opening (same as me) followed by a char that is not a hash -- that then gets thrown away (`_2` is never concated) then it allows the same subset of chars mine does, with a similar concat to mine.

So, to my mind, yours has a confusing Regex at the start, which fills an unclear purpose, and then yours has a seemingly too wide `Char().Not()` that then gets thrown away too, and yet somehow seems to work regardless. To my mind, your approach is imply 1. less clear and 2. less consistent with how the other parsers work.

To your points, 1 would be very nice, tho I'm not sure I'm seeing that, and I would like you to point at the specific tests for 2 and 3 that you think are compelling that are not present in my PR -- nor even before either PR.

Reply:
OK, I added the regex to the start of mine https://github.com/tonerdo/dotnet-env/pull/87/commits/dc277bebea97f0378ab2f96eb05e3656db326dac to understand what you were talking about, and I see now what you mean with the error messages being from unquoted values matching empty and then leaving unconsumed extra chars, vs quoted values matching but missing the final quote. That is an improvement, for sure.

Reply:
I am still unclear what you mean by 2 and 3.

For 2, I did review my own new tests that fail with the previous code: https://github.com/tonerdo/dotnet-env/pull/87#issuecomment-1870680310 and that was already the case before, if I understand you correctly, and still is. Meaning, I am not clear what you are arguing about those tests are unique about your approach for `UnquotedValue`

For 3, I believe at this point we both have all the same tests (which we should) -- I only did not have your changes to the error messages, which I added once I understood what you meant and were aiming for.

Reply:
At this point, I like your UnquotedValue + UnquotedValueContents approach with from-select. I certainly want you to get the credit for the changes to the repo, as this is instigated by you :)

I do ask that you make sure you are including all the new tests I have (I believe we are all the same now, I just want you to double check). I do want you to use the exact error message tests I added to my PR tho, with the comment above, and the `abc` after the quote char to make the position of the failure more explicit -- and also the comment I added for the regex in my PR into your from-select approach as well, please.

Reply:
Oh, and to be clear: please replace `Parse.Regex("...").Not()` with simply `Parse.Chars(" \t\"'").Not()` as that remains cleaner for my tastes

Reply:
First, thank you for giving me the credit for all of this stuff! It's quite some of your work and reviews too.

I will dive through all of it and work at the last todos.

Great that you found the reason behind the leading Not for the ErrorMessage on your own, I should have clarified that more precise and showing the exact test-case as you said. It's quite hard to find the correct "layer" of description if it's all so clear in the own head :D

About the Not(): In general I agree, that is definitely often confusing and that's the main reason why I was not really happy with the "Then" solution.
But with the "select-from" I think it is quite good readable and understandable now, it's just an initial check that we do not start with certain characters (yes. I will change it to Chars ;) ).
The reason why I would not try to move this checks inside Interpolated or the "default" ValueContent is about "separation of concerns" I would say. There is a single Statement now which says: "the beginning character is never allowed to be that". And if that is checked, we can go on with parsing the actual value.

It would have to be tested, but I think it is possible to to an `Except` at the end of the Parser instead of the `Not` in front of it, but I would prefer the solution with Not, because it avoids to think in the "wrong order" (it's a little bit like from inside to outside instead of fluent, but within fluent, so from behind to front there...)
The unused `_` and `_2` is just an empty object, because `Not` will never consume and return anything.

I guess I will post the `Except` solution afterwards for reference.

Reply:
Now I see again why I would prefer my version with the Not for the value itself. I really focused on the outer one only before.

It's more or less equal to each other and I'm not sure which one to use now. Not sure if you like both from-selects or just the outer one ;)
I leave it with mine, including the comment for now waiting for your response.

Here is the reason:
With the `Not` we separate the check for the first character from the general character-check.
Means with my solution there is a single check for the `#` only, which doesn't consume anything. And then every character gets checked with `NotControlNorWhitespace("$\"'")`.
Your solution on the other hand checks and consumes the first character in a single step, so you must assure that the list of additionally excluded characters is "in sync" between both calls, except for the hash, which is relevant for the first character only.

Reply:
For Quoted Values there is some separation of tests between Contents and Wrapper, should I do this too or should we stick with the current tests?

Reply:
I am fine with the Not() now that I understand that it is non consuming.

I don't feel a need for separating Content vs wrapper tests at this point. Let's just leave the tests we have now (again, making sure yours and mine are unified).

Reply:
Now for reference, this is working too:
```
        internal static readonly Parser<ValueCalculator> UnquotedValue =
            UnquotedValueContents.Except(Parse.Chars(" \t\"'"));
```
Having the separation of UnquotedValue and UnquotedValueContent is is okay I would say, but I really dislike this order of statements, but it's kinda the same with Not I guess, which is put at the last position...
But with the separation from `select-from` this is quite explanative imho.